File: app/src/main/java/org/wikipedia/settings/languages/WikipediaLanguagesItemView.java
package org.wikipedia.settings.languages;

import android.content.Context;
import android.graphics.PorterDuff;
import android.os.Build;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.content.res.AppCompatResources;

import org.apache.commons.lang3.StringUtils;
import org.wikipedia.R;
import org.wikipedia.databinding.ItemWikipediaLanguageBinding;
import org.wikipedia.util.ResourceUtil;
import org.wikipedia.views.ViewUtil;

import static org.wikipedia.search.SearchFragment.LANG_BUTTON_TEXT_SIZE_LARGER;
import static org.wikipedia.search.SearchFragment.LANG_BUTTON_TEXT_SIZE_SMALLER;

public class WikipediaLanguagesItemView extends LinearLayout {
    public interface Callback {
        void onCheckedChanged(int position);
        void onLongPress(int position);
    }

    private TextView orderView;
    private CheckBox checkBox;
    private TextView titleView;
    private TextView langCodeView;
    private View dragHandleView;
    @Nullable private Callback callback;
    private int position;

    public WikipediaLanguagesItemView(Context context) {
        super(context);
        init();
    }

    public WikipediaLanguagesItemView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public WikipediaLanguagesItemView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    public void setCallback(@Nullable Callback callback) {
        this.callback = callback;
    }

    public void setContents(@NonNull String langCode, @Nullable String languageLocalizedName, int position) {
        this.position = position;
        orderView.setText(String.valueOf(position + 1));
        titleView.setText(StringUtils.capitalize(languageLocalizedName));
        ViewUtil.formatLangButton(langCodeView, langCode, LANG_BUTTON_TEXT_SIZE_SMALLER, LANG_BUTTON_TEXT_SIZE_LARGER);
        langCodeView.setText(langCode);
        langCodeView.setTextColor(ResourceUtil.getThemedColor(getContext(), R.attr.material_theme_de_emphasised_color));
        langCodeView.getBackground().setColorFilter(ResourceUtil.getThemedColor(getContext(), R.attr.material_theme_de_emphasised_color),
                PorterDuff.Mode.SRC_IN);
    }

    public void setCheckBoxEnabled(boolean enabled) {
        orderView.setVisibility(enabled ? GONE : VISIBLE);
        checkBox.setVisibility(enabled ? VISIBLE : GONE);
        if (!enabled) {
            checkBox.setChecked(false);
            setBackgroundColor(ResourceUtil.getThemedColor(getContext(), R.attr.paper_color));
        }
    }

    public void setCheckBoxChecked(boolean checked) {
        checkBox.setChecked(checked);
        updateBackgroundColor();
    }

    public void setDragHandleEnabled(boolean enabled) {
        dragHandleView.setVisibility(enabled ? VISIBLE : GONE);
    }

    public void setDragHandleTouchListener(OnTouchListener listener) {
        dragHandleView.setOnTouchListener(listener);
    }

    private void init() {
        final ItemWikipediaLanguageBinding binding = ItemWikipediaLanguageBinding.bind(this);

        orderView = binding.wikiLanguageOrder;
        checkBox = binding.wikiLanguageCheckBox;
        titleView = binding.wikiLanguageTitle;
        langCodeView = binding.wikiLanguageCode;
        dragHandleView = binding.wikiLanguageDragHandle;

        binding.getRoot().setOnLongClickListener(v -> {
            if (callback != null) {
                callback.onLongPress(position);
            }
            return true;
        });
        checkBox.setOnCheckedChangeListener((v, isChecked) -> {
            if (callback != null) {
                callback.onCheckedChanged(position);
                updateBackgroundColor();
            }
        });

        setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT));
        setBackgroundColor(ResourceUtil.getThemedColor(getContext(), R.attr.paper_color));
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            setForeground(AppCompatResources.getDrawable(getContext(),
                    ResourceUtil.getThemedAttributeId(getContext(), R.attr.selectableItemBackground)));
        }
    }

    private void updateBackgroundColor() {
        setBackgroundColor(checkBox.isChecked()
                ? ResourceUtil.getThemedColor(getContext(), R.attr.multi_select_background_color) : ResourceUtil.getThemedColor(getContext(), R.attr.paper_color));
    }
}

File: app/src/main/java/org/wikipedia/settings/languages/WikipediaLanguagesItemView.kt
package org.wikipedia.settings.languages

import android.annotation.SuppressLint
import android.content.Context
import android.os.Build
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.view.ViewCompat
import org.wikipedia.R
import org.wikipedia.databinding.ItemWikipediaLanguageBinding
import org.wikipedia.language.LanguageUtil
import org.wikipedia.search.SearchFragment
import org.wikipedia.util.DeviceUtil
import org.wikipedia.util.ResourceUtil
import org.wikipedia.util.StringUtil
import org.wikipedia.views.ViewUtil
import java.util.*

class WikipediaLanguagesItemView : LinearLayout {
    interface Callback {
        fun onCheckedChanged(position: Int)
        fun onLongPress(position: Int)
    }

    private var binding = ItemWikipediaLanguageBinding.inflate(LayoutInflater.from(context), this)
    private var position = 0
    var callback: Callback? = null

    constructor(context: Context?) : super(context)
    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    init {
        layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT)
        setBackgroundColor(ResourceUtil.getThemedColor(context, R.attr.paper_color))
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            foreground = AppCompatResources.getDrawable(context,
                ResourceUtil.getThemedAttributeId(context, androidx.appcompat.R.attr.selectableItemBackground))
        }
        binding.wikiLanguageCheckbox.setOnCheckedChangeListener { _, _ ->
            callback?.onCheckedChanged(position)
            updateBackgroundColor()
        }
        setOnLongClickListener {
            callback?.onLongPress(position)
            true
        }
        DeviceUtil.setContextClickAsLongClick(this)
    }

    private fun updateBackgroundColor() {
        setBackgroundColor(if (binding.wikiLanguageCheckbox.isChecked) ResourceUtil.getThemedColor(context, R.attr.background_color)
        else ResourceUtil.getThemedColor(context, R.attr.paper_color))
    }

    fun setContents(langCode: String, languageLocalizedName: String?, position: Int) {
        this.position = position
        binding.wikiLanguageOrder.text = (position + 1).toString()
        binding.wikiLanguageTitle.text = StringUtil.capitalize(languageLocalizedName.orEmpty())
        binding.wikiLanguageCode.text = LanguageUtil.formatLangCodeForButton(langCode)
        val color = ResourceUtil.getThemedColorStateList(context, R.attr.secondary_color)
        binding.wikiLanguageCode.setTextColor(color)
        ViewCompat.setBackgroundTintList(binding.wikiLanguageCode, color)
        ViewUtil.formatLangButton(binding.wikiLanguageCode, langCode, SearchFragment.LANG_BUTTON_TEXT_SIZE_SMALLER, SearchFragment.LANG_BUTTON_TEXT_SIZE_LARGER)
    }

    fun setCheckBoxEnabled(enabled: Boolean) {
        binding.wikiLanguageOrder.visibility = if (enabled) GONE else VISIBLE
        binding.wikiLanguageCheckbox.visibility = if (enabled) VISIBLE else GONE
        if (!enabled) {
            binding.wikiLanguageCheckbox.isChecked = false
            setBackgroundColor(ResourceUtil.getThemedColor(context, R.attr.paper_color))
        }
    }

    fun setCheckBoxChecked(checked: Boolean) {
        binding.wikiLanguageCheckbox.isChecked = checked
        updateBackgroundColor()
    }

    fun setDragHandleEnabled(enabled: Boolean) {
        binding.wikiLanguageDragHandle.visibility = if (enabled) VISIBLE else GONE
    }

    @SuppressLint("ClickableViewAccessibility")
    fun setDragHandleTouchListener(listener: OnTouchListener?) {
        binding.wikiLanguageDragHandle.setOnTouchListener(listener)
    }
}
